contents/why_julia.md

```diff
@@ -133,8 +136,20 @@ We don't know any other serious language that would want to be hard and slow, so
 
 **Julia is fast!
 Very fast!**
-It was made for speed from the drawing board.
-It accomplishes this by exposing native machine code to the compiler.
+It was designed for speed from the beginning.
+It accomplishes this by multiple dispatch.
+Basically, the idea is to generate very efficient LLVM code.
+For example, if you define a function taking one argument and pass an integer into the function, then Julia will create a _specialized_ `MethodInstance`.
+The next time that you pass an integer to the function, Julia will lookup the `MethodInstance` that was created earlier and refer execution to that.
+Now, the **great** trick is that you can also do this inside a function that calls a function.
+For example, if some data type is passed into function `f` and `f` calls function `g` and the data types passed to `g` are known and always the same, then the generated function `g` can be hardcoded into function `f`!
+This means that Julia doesn't even have to lookup `MethodInstances` anymore, and the code can run very efficiently.
+The tradeoff, here, is that there are cases where earlier assumptions about the hardcoded `MethodInstances` are invalidated.
+Then, the `MethodInstance` has to be recreated which takes time.
+Also, the tradeoff is that it takes time to infer what can be hardcoded and what not.
+This explains why it can often take very long before Julia does the first thing:
+in the background, it is optimizing your code.
```

Comment:
I guess that you maybe get this explanation, but people new to Julia do not. Do you have tips on how I could make this explanation more clear, @storopoli?

Reply:
People don't know what LLVM code is. I would prefer to "dummyfy" a little bit and just say machine code. Or we could add something like

> very efficient LLVM code which translates to machine code.

Reply:
Okay. I've added an explanation. Is it clear enough now, you think?

Reply:
Yes!